### DISCUSSION TITLE
Relazione tra i contatori utility meter e il database home-assistant_v2.db

### DISCUSSION THREAD

**Discussion by Newreader** [indomus/forum#2214] [HUB: Home Assistant]
Salve ,
vorrei porre una mia personale riflessione in merito agli utility meter (modificato rispetto alla prima stesura).
Così come definiti in HA I sensori creati con l’integrazione utility meter sono persistenti, quindi i valori vengono mantenuti durante i riavvii di Home Assistant, questo vuol dire che vengono memorizzati nel database home-assistant_v2.db, conseguentemente se ad esempio volessi calcolare il valore di un contatore con una frequenza mensile o annuale non avrei mai il valore corretto se ponessi un purge di un tempo inferiore , ad esempio 7 giorni.
Quindi per poter contabilizzare correttamente tutti i contatori che superano il limite temporale esposto in purge (esempio 7 giorni), è necessario innalzare tale limite alla massima frequenza, ovvero mensile o annuale, di registrazione (con la problematica dei giorni dei diversi mesi e/o degli anni bisestili che tuttavia si possono risolvere con automazioni ad hoc).
Questo comporta che, comunque, occorre portarsi dietro tutti quei sensori che vogliamo valorizzare a fini statistici anche se hanno una cadenza (ad esempio settimanale o mensile nel caso di cronologia pari a 365/366 giorni). 
Ringrazio chiunque volesse darmi chiarimenti in merito e suggerirmi strade alternative, a prescindere da InfluxDB, DBMaria, Grafana e simili

**Comment by KC-inDomus** [comment]
> Salve , vorrei porre una mia personale riflessione in merito agli utility meter (modificato rispetto alla prima stesura). Così come definiti in HA I sensori creati con l’integrazione utility meter sono persistenti, quindi i valori vengono mantenuti durante i riavvii di Home Assistant, questo vuol dire che vengono memorizzati nel database home-assistant_v2.db, conseguentemente se ad esempio volessi calcolare il valore di un contatore con una frequenza mensile o annuale non avrei mai il valore corretto se ponessi un purge di un tempo inferiore , ad esempio 7 giorni. Quindi per poter contabilizzare correttamente tutti i contatori che superano il limite temporale esposto in purge (esempio 7 giorni), è necessario innalzare tale limite alla massima frequenza, ovvero mensile o annuale, di registrazione (con la problematica dei giorni dei diversi mesi e/o degli anni bisestili che tuttavia si possono risolvere con automazioni ad hoc). Questo comporta che, comunque, occorre portarsi dietro tutti quei sensori che vogliamo valorizzare a fini statistici anche se hanno una cadenza (ad esempio settimanale o mensile nel caso di cronologia pari a 365/366 giorni). Ringrazio chiunque volesse darmi chiarimenti in merito e suggerirmi strade alternative, a prescindere da InfluxDB, DBMaria, Grafana e simili

io ho un purge impostato a 8 giorni, ma gli utility li ho con profondità di mesi e mesi. non mi risulta che il recorder abbia un impatto su questo tipo di entità.

**Reply by Newreader** [reply]
Grazie per la risposta.
Comunque mi permetto di dissentire per due motivazioni una esperienziale e una logica (almeno per me). 
Uso uno Shelly EM (configurato per MQTT) montato a monte dell'impianto per poter monitorare i consumi e poter gestire i surplus di richiesta di potenza con delle automazioni personali onde evitare il distacco di corrente.
Come è noto lo Shelly EM ha alcuni contatori che misurano quanto preso dalla rete uno fa riferimento a "energy" (contatore di energia in Watt-minuto) e un altro fa riferimento a "total" energia totale accumulata in Wh, che diversamente dal primo che non sopravvive  al ciclo di accensione o al riavvio, viene memorizzato nella memoria non volatile del dispositivo.
Preciso che io ho un purge impostato a 7.
Ciò premesso, se faccio partire  i due contabilizzatori  "total" e utility meter (yearly, i cui valori sono persistenti e quindi mantenuti durante i riavvii di Home Assistant ) da un valore iniziale pari a zero dovrei avere in entrambi lo stesso valori (a meno di piccole differenze dovute al diverso metodo di calcolo, forse). Ebbene questo non si verifica, nel contabilizzatore derivato da utility_meter (yearly), ho un valore molto più basso che in quello "total" gestito direttamente dallo Shelly EM.
Altra verifica (empirica) che ho effettuato è quella di salvarmi i dati giornalieri del consumo (utility meter daily, ovviamente sincroni con gli altri valori summenzionati) attraverso un messaggio che mi viene inviato a Telegram a mezzanotte di ogni giorno (anzi per la precisione alle 23:59:59), se sommo quindi i consumi dall'inizio del conteggio sia per i dati mensili che per i dati annuali (con riferimento al 2022) non ho assolutamente il riscontro/concordanza dei dati rilevati (monthly, yearly) ma piuttosto differenze sostanziali (in questo caso sto usando solo entità utility meter).
La motivazione logica nasce da questa considerazione (già peraltro enucleata nel post iniziale: i valori di queste entità utility meter per poter essere disponibili devono essere salvati da qualche parte e, penso che sia proprio il file home-assistant_v2.db deputato a farlo, Tant'è che se si cancella questo db si perde, come è risaputo, lo storico.
Ora una cosa non mi è nota: queste entità che presumo vengano salvate singolarmente, es. daily, monthly, yearly, non subiscono l'effetto del purge per impostazione? Non ho trovato, purtroppo, alcuna informazione in merito.
Mi sono ripromesso di fare altre verifiche ed altri test per approfondire l'argomento e cercare di venirne a capo (anche perché ne va della bontà del monitoraggio fatto utilizzando gli utility meter).
Spero di continuare ad avere, in questa sede, altre tue considerazioni in merito.
